CIPs/0017.md

```diff
@@ -0,0 +1,181 @@
+```
+GetProofInventory - Asks the server what proofs it has. Server responds with `[]proofMetadata` of all available proofs.
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, handleLogic boolean, requestedProofs []proofMetadata)
+    `highestEpoch` is the current highestEpoch the client knows of.
+    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
+                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
+    `requestedProofs`: Set of proofs' metadata requested by the client
+  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
+  Both responses are subject to message size limits limiting the number of proofs/headers.
+EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.
+```
+
+The client will call `GetProofInventory`, either during the sync or pipelined within the handshake. This information will then be stored for each peer, and before each request the client will assess the "best peer", where "best" here would be defined as the minimum number of proofs & headers needed to sync from the client's epoch to `currEpoch`. This calculation could work as follows: 
+
+`numEpochs`: Number of epochs we need to sync, the difference between the current epoch of the canonical chain (`currEpoch`) and the client's highest epoch `clientEpoch` = `currEpoch - clientEpoch`
+
+`optimalPath`: For each peer, compute the optimal path of proofs available based on the client's available VKs and known val sets.
+
+`proofRange` : For each proof, the difference between `LastEpoch` and `FirstEpoch`'s indices.
+
+`peerScore`: This is the total number of proofs and headers needed to sync by this peer. For each peer, subtract the range of each proof in `optimalPath` from `numEpochs`, the result is the number of necessary remaining headers. Add `optimalPath`.length to this for the peer's score. 
+
+Whichever peer has the lowest score indicates the shortest path (best set of proofs) for our client. This could also be short-circuited if the optimal score is provided. Once the best peer is chosen, the client requests the set of proofs and headers from this peer that it has assessed as optimal.
+
+Option 2:
+```
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, knownValSets *big.Int, maxResponses uint)
+    `knownValSets` is a bitmap of the last ~240 or so epochs that the client knows the val set for
```

Comment:
I mentioned before, but we should do some thinking about when a bitmap is more or less efficient than a range based representation (i.e. sending a list of `[start, end]` `uint32` pairs that represent that ranges of epochs the light client knows the validator sets for.)

Reply:
Agreed I'll think more about this and update now that it seems this is the decided (in my mind) path forward